### PR TITLE
Golang implementation of Time Series Merge

### PR DESCRIPTION
**Pull request by mrtracy**
This commit creates a Golang implementation of the logic to merge
InternalTimeSeriesData messages together; this has also been implemented in C++
as part of our RocksDB custom merge operator.

The Go implementation is primarily intended for testing purposes - it provides
the ability to double-check the logic present in the C++ engine as part of high
level, model based testing which is currently in development. The Golang
implementation is run through the exact same merge tests as the engine
implementation.

As part of this commit, I have also repaired a bug in the C++ engine merge logic
; previously, it was not properly sorting and accumulating the samples from the
first merge operation on a given key, instead just accepting the collection
exactly as it was provided.

### CODE REVIEW

File: storage/engine/db.cc [hunk 1]
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
...
+
+    // Initialize new_ts and its primitive data fields.
+    cockroach::proto::InternalTimeSeriesData new_ts;
+    new_ts.set_start_timestamp_nanos(val_ts.start_timestamp_nanos());
+    new_ts.set_sample_duration_nanos(val_ts.sample_duration_nanos());
+
+    // Sort values in the ts value. 
+    std::sort(val_ts.mutable_samples()->pointer_begin(),
+            val_ts.mutable_samples()->pointer_end(),
+            TimeSeriesSampleOrdering);
```
Comment (petermattis) [inline storage/engine/db.cc]: Looks like these two lines are mis-indented.

File: storage/engine/db.cc [hunk 2]
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
...
+
+    // Merge sample values of left and right into new_ts.
+    auto front = val_ts.samples().begin(),
+         end = val_ts.samples().end();
+
+    // Loop until samples have been exhausted.
+    while(front != end) {
```
Comment (petermattis) [inline storage/engine/db.cc]: `s/while(/while (/g`

File: storage/engine/db.cc [hunk 3]
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
...
+    new_ts.set_sample_duration_nanos(val_ts.sample_duration_nanos());
+
+    // Sort values in the ts value. 
+    std::sort(val_ts.mutable_samples()->pointer_begin(),
+            val_ts.mutable_samples()->pointer_end(),
+            TimeSeriesSampleOrdering);
+
+    // Merge sample values of left and right into new_ts.
+    auto front = val_ts.samples().begin(),
+         end = val_ts.samples().end();
```
Comment (petermattis) [inline storage/engine/db.cc]: In C++ I prefer to see 1 variable definition per line and to avoid using commas to separate multiple variable declarations. Pretty sure this is part of the Google C++ coding guidelines which we generally follow.

File: storage/engine/db.cc [hunk 4]
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
...
+    while(front != end) {
+        // Create an empty sample in the output collection with the selected
+        // offset.  Accumulate data from all samples at the front of the sample
+        // collection which match the selected timestamp. This behavior is
+        // needed because even a single value may have duplicated offsets. 
+        cockroach::proto::InternalTimeSeriesSample* ns = new_ts.add_samples();
+        ns->set_offset(front->offset());
+        while (front != end && front->offset() == ns->offset()) {
+            AccumulateTimeSeriesSamples(ns, *front);
+            front++;
```
Comment (petermattis) [inline storage/engine/db.cc]: With C++ iterators you should prefer pre-increment to post-increment: `s/front++/++front/g`

File: storage/engine/db.cc (hunk 1 above)
Comment (petermattis) [inline storage/engine/db.cc]: Not part of this change, but you should use a functor object instead of a function pointer for `TimeSeriesSampleOrdering`. In my past experience functor objects inline better (might not be true of recent compilers). Something like:

```
struct TimeSeriesSampleOrdering {
  bool operator()(const cockroach::proto::InternalTimeSeriesSample* a, const cockroach::proto::InternalTimeSeriesSample* b) const {
    return a->offset() < b->offset();
  }
};
```

File: storage/engine/merge_test.go [hunk 5]
```diff
@@ -353,20 +365,51 @@ func TestGoMerge(t *testing.T) {
...
+
+// unmarshalTimeSeries unmarshals the time series value stored in the given byte
+// array. It is assumed that the time series value was originally marshalled as
+// a proto.MVCCMetadata with an inline value.
+func unmarshalTimeSeries(t testing.TB, b []byte) *proto.InternalTimeSeriesData {
+	if b == nil {
+		return nil
+	}
+	var mvccValue proto.MVCCMetadata
+	gogoproto.Unmarshal(b, &mvccValue)
```
Comment (petermattis) [inline storage/engine/merge_test.go]: Doesn't `gogoproto.Unmarshal` return an error? You should be checking it.

File: storage/engine/db.cc (hunk 4 above)
Comment (mrtracy) [inline storage/engine/db.cc]: Any reason for that?  Style guide, or an implementation advantage?

File: storage/engine/db.cc (hunk 3 above)
Comment (tnachen) [inline storage/engine/db.cc]: Yes Google C++ style states so too.

File: storage/engine/merge_test.go (hunk 5 above)
Comment (mrtracy) [inline storage/engine/merge_test.go]: Done

File: storage/engine/db.cc (hunk 3 above)
Comment (mrtracy) [inline storage/engine/db.cc]: Done

File: storage/engine/db.cc (hunk 2 above)
Comment (mrtracy) [inline storage/engine/db.cc]: Done

File: storage/engine/db.cc (hunk 4 above)
Comment (petermattis) [inline storage/engine/db.cc]: It's an implementation advantage. The post-increment requires a copy. The pre-increment does not. See the difference between https://code.google.com/p/cpp-btree/source/browse/btree.h#812 (pre-increment) and https://code.google.com/p/cpp-btree/source/browse/btree.h#820 (post-increment).

File: proto/internal.go [hunk 6]
```diff
@@ -87,3 +91,190 @@ func InternalTimeSeriesDataFromValue(value *Value) (*InternalTimeSeriesData, err
 	}
 	return &ts, nil
 }
+
+// AccumulateFrom accumulates values from the supplied InternalTimeSeriesSample
+// into the existing values of the destination InternalTimeSeriesSample.
+func (dest *InternalTimeSeriesSample) AccumulateFrom(src *InternalTimeSeriesSample) {
```
Comment (petermattis) [inline proto/internal.go]: Any reason for this to be part of the `proto` package. If it is only used by `storage/engine/merge_test.go`, I'd prefer for it to be private to the `storage/engine` package.

Comment (mrtracy) [inline proto/internal.go]: It's not only going to be used there, it's also going to support testing in the upcoming /ts module.  That's the primary reason for this PR.

**Comment by petermattis** [comment]
LGTM

File: proto/internal.go [hunk 7]
```diff
@@ -87,3 +91,190 @@ func InternalTimeSeriesDataFromValue(value *Value) (*InternalTimeSeriesData, err
...
+		// can be nil in the original messages, in which case the Sum is both the
+		// min and the max.
+		valOrDefault := func(maybeVal *int64, defaultVal int64) int64 {
+			if maybeVal != nil {
+				return *maybeVal
+			}
+			return defaultVal
+		}
+		var (
+			destMax = valOrDefault(dest.IntMax, dest.GetIntSum())
```
Comment (spencerkimball) [inline proto/internal.go]: Looking at this, I'm wondering whether you think it might be overkill to support both Ints and Floats. I know we discussed this before but I don't recall the convincing argument that we support both. I'd be tempted to just support Float64. I'm worried about all of these allocations. hmm.

**Comment by spencerkimball** [comment]
I'm not a big fan of implementing the same logic twice. If we only need the go logic in order to do testing, let's just create a cgo method which allows the C++ merging code to be called from Go. You can just pass the cgo method the serialized protobufs and deserialize on the C++ side, then run through the C++ merge, serialize again and then deserialize the result on the Go side.

Maybe I'm missing something...

**Comment by mrtracy** [comment]
The C++ Merging code can already be called from go (it's needed for batch),
but it's not being exported from the engine module.  I need this for the
model-based testing i'm using in the /ts module, for which I haven't yet
sent out a PR.

I made the decision that re-implementing the logic in Go would be
beneficlal, as it is a fairly complicated transformation.  The real
implementation is an optimized C++ implementation which is called in the
context of a RocksDB Merge Operator, and so it's not especially
straightforward to understand.  I feel that this is a solid way to double
check the implementation in more complicated situations.

If you feel strongly about having only one implementation, I can export the
rename and export the goMerge method from the engine layer.

**Comment by spencerkimball** [comment]
I actually do feel pretty strongly that we shouldn't duplicate this code. It's great that it allowed you to find a bug, but I think ultimately having two implementations will lead to more confusion and maintenance than it's worth. If you want to use it for model based testing in the /ts module, let's just provide some exported methods from engine which call into the C++.

**Comment by mrtracy** [comment]
I have removed the go implementation (which significantly changes the purpose of this commit) and replaced it with an exported helper method in the engine, which directly provides for the merging of InternalTimeSeriesData objects.

File: storage/engine/db.cc [hunk 8]
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
     return true;
 }
 
+// ConsolidateTimeSeriesValue processes a single value which contain
```
Comment (spencerkimball) [inline storage/engine/db.cc]: s/contain/contains/

**Comment by spencerkimball** [comment]
LGTM
